app/main/views/service_settings.py

```diff
@@ -402,19 +405,100 @@ def service_email_reply_to(service_id):
+        ) + "?is_default={}".format(is_default)
         )
-        return redirect(url_for('.service_email_reply_to', service_id=service_id))
+
     return render_template(
         'views/service-settings/email-reply-to/add.html',
         form=form,
         first_email_address=first_email_address)
 
 
+@main.route("/services/<service_id>/service-settings/email-reply-to/<notification_id>/verify", methods=['GET', 'POST'])
+@login_required
+@user_has_permissions('manage_service')
+def service_verify_reply_to_address(service_id, notification_id):
+    replace = request.args.get('replace', False)
+    request_args = "?is_default={}&replace={}".format(request.args.get('is_default', False), replace)
+    return render_template(
+        'views/service-settings/email-reply-to/verify.html',
+        service_id=service_id,
+        notification_id=notification_id,
+        partials=get_service_verify_reply_to_address_partials(service_id, notification_id),
+        verb=("Change" if replace else "Add"),
+        replace=replace,
+        request_args=request_args
+    )
+
+
+@main.route("/services/<service_id>/service-settings/email-reply-to/<notification_id>/verify.json")
+@login_required
+@user_has_permissions('manage_service')
+def service_verify_reply_to_address_updates(service_id, notification_id):
+    return jsonify(**get_service_verify_reply_to_address_partials(service_id, notification_id))
+
+
+def get_service_verify_reply_to_address_partials(service_id, notification_id):
+    form = ServiceReplyToEmailForm()
+    first_email_address = current_service.count_email_reply_to_addresses == 0
+    notification = notification_api_client.get_notification(current_app.config["NOTIFY_SERVICE_ID"], notification_id)
+    replace = request.args.get('replace', False)
+    verification_status = "pending"
+    is_default = True if (request.args.get('is_default', False) == "True") else False
+    if notification["status"] == "delivered":
+        verification_status = "success"
+        if notification["to"] not in [i["email_address"] for i in current_service.email_reply_to_addresses]:
+            if replace and replace != "False":
+                service_api_client.update_reply_to_email_address(
+                    current_service.id, replace, email_address=notification["to"], is_default=is_default
+                )
+            else:
+                service_api_client.add_reply_to_email_address(
+                    current_service.id,
+                    email_address=notification["to"],
+                    is_default=is_default
+                )
+    created_at_no_tz = notification["created_at"][:-6]
+    seconds_since_sending = (datetime.utcnow() - datetime.strptime(created_at_no_tz, '%Y-%m-%dT%H:%M:%S.%f')).seconds
+    if notification["status"] in [
+        "failed", "permanent-failure", "technical-failure", "temporary-failure"
+    ] or (notification["status"] == "sending" and seconds_since_sending > 300):
```

Comment:
I think 90 seconds is better. It’s close to ‘a minute’, which is what we say on the page, but with a bit of breathing room.
```suggestion
    ] or (notification["status"] == "sending" and seconds_since_sending > 90):
```

Reply:
👍

Reply:
how long does it usually take for us to get a delivery receipt?

Reply:
I did some analysis of a random sample of email notifications in the `notifications` table.

– | Time to get ‘delivered‘ status (seconds) | Time to get ‘failure‘ status (seconds)
---|---|---
Mean | 11.6 | 22,142.1
Median | 2.9 | 8.7

In percentile terms it looks like this:

![image](https://user-images.githubusercontent.com/355079/58110437-6db8cc00-7be7-11e9-8e03-c3d267f04630.png)

You can see that failure is slower than success. So the longer a notification takes to get a status, the more likely it is for that status to be a failure anyway. This increases dramatically after 45 seconds. The percentage of emails that go to `delivered` in less than 90 seconds is 98.92%. To get to 99% we’d need to increase the timeout to 178 seconds (3 minutes). We could still get 98.7% of notifications by dropping the timeout to 45 seconds, and improve the experience for notifications that are likely to fail by returning an error more quickly.

Source: https://docs.google.com/spreadsheets/d/1oWttUHbT1eWzzvljeFu0w7VbekD7dlRbjNyYiMj68QU/edit#gid=1850531376

Reply:
nice one with the analysis, I updated it now and included part of your above comment in the commit message for future generations :P